openSearchLane.go

```diff
@@ -0,0 +1,469 @@
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	if osl.metadata == nil {
+		osl.metadata = make(map[string]string)
+	}
+
+	osl.metadata[key] = val
+}
+
+func (osl *openSearchLane) Close() {
+	osl.openSearchConnection.detach()
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	if strings.Contains(logEntry, "\n") {
+		logEntry = strings.ReplaceAll(logEntry, "\n", "")
+	}
+
+	// TODO remove this line
+	fmt.Println(logEntry)
+
+	parentLaneId, _ := osl.Context.Value(parent_lane_id).(string)
+
+	if osl.metadata != nil {
+		osl.metadata["timestamp"] = time.Now().UTC().Format(time.RFC3339)
+	}
+
+	logData := &OpenSearchLogMessage{
+		AppName:      osl.openSearchConnection.config.OpenSearchAppName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+		Metadata:     osl.metadata,
+	}
+
+	osl.openSearchConnection.logBuffer = append(osl.openSearchConnection.logBuffer, logData)
+
+	osl.openSearchConnection.messagesQueued++
+
+	if ((1 + osl.openSearchConnection.messagesSent - osl.openSearchConnection.messagesQueued) % osl.openSearchConnection.config.LogThreshold) == 0 {
+		osl.openSearchConnection.wakeCh <- false
+	}
+
+	return len(p), nil
+}
+
+func (osl *openSearchLane) Stats() (stats OslStats) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	stats.MessagesQueued = osl.openSearchConnection.messagesQueued
+	stats.MessagesSent = osl.openSearchConnection.messagesSent
+	stats.MessagesSentFailed = osl.openSearchConnection.messagesSentFailed
+
+	return
+}
+
+func (osl *openSearchLane) SetEmergencyHandler(emergencyFn OslEmergencyFn) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+	osl.openSearchConnection.emergencyFn = emergencyFn
+}
+
+func (osc *openSearchConnection) processConnection() {
+	backoffDuration := osc.config.BackoffInterval
+
+	for {
+		select {
+		case wg := <-osc.stopCh:
+			shouldStop := false
+			osc.mu.Lock()
+			osc.refCount--
+			shouldStop = (osc.refCount == 0)
+			osc.mu.Unlock()
+
+			if shouldStop {
+				go func() {
+					osc.mu.Lock()
+					if len(osc.logBuffer) > 0 {
+						err := osc.flush(osc.logBuffer)
+						if err != nil {
+							if osc.emergencyFn != nil {
+								err = osc.emergencyFn(osc.logBuffer)
+								if err != nil {
+									osc.messagesSentFailed += len(osc.logBuffer)
+								} else {
+									osc.messagesSent += len(osc.logBuffer)
+								}
+							}
+						}
+						osc.logBuffer = make([]*OpenSearchLogMessage, 0)
+					}
+					osc.mu.Unlock()
+					wg.Done()
+				}()
+				return
+			}
+
+			wg.Done()
+
+		case isClientActive := <-osc.wakeCh:
+			if isClientActive {
+				backoffDuration = osc.send(backoffDuration)
+				osc.client = nil
+			} else {
```

Comment:
Add synchronization that this got done (wg) to connect() doesn't clobber the config that another case here quickly picks up before the connect() go routine resumes